### PR TITLE
Removed invalid reference usage in charuco detector

### PR DESCRIPTION
**Pull request by asmorkalov**
Address https://github.com/opencv/opencv/commit/60ae97314236c5c6d96d15a052a6dc2a1fdc3846#r127863372

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/objdetect/src/aruco/charuco_detector.cpp [hunk 1]
```diff
@@ -27,10 +27,9 @@ struct CharucoDetector::CharucoDetectorImpl {
...
         vector<Mat> mCorners;
         markerCorners.getMatVector(mCorners);
-        const Mat& mIds = markerIds.getMat();
-
-        const Mat& chCorners = charucoCorners.getMat();
-        const Mat& chIds = charucoIds.getMat();
+        Mat mIds = markerIds.getMat();
+        Mat chCorners = charucoCorners.getMat();
+        Mat chIds = charucoIds.getMat();
         const vector<int>& boardIds = board.getIds();
```
Comment (vrabaud) [inline modules/objdetect/src/aruco/charuco_detector.cpp:33]: getIds() returns a `const vector<int>&` so this is fine

**Review by vrabaud — CHANGES_REQUESTED** [review]
Please also fix the `nearestMarkerIdx` reference.

File: modules/objdetect/src/aruco/charuco_detector.cpp [hunk 2]
```diff
@@ -27,10 +27,9 @@ struct CharucoDetector::CharucoDetectorImpl {
     bool checkBoard(InputArrayOfArrays markerCorners, InputArray markerIds, InputArray charucoCorners, InputArray charucoIds) {
         vector<Mat> mCorners;
         markerCorners.getMatVector(mCorners);
-        const Mat& mIds = markerIds.getMat();
-
-        const Mat& chCorners = charucoCorners.getMat();
-        const Mat& chIds = charucoIds.getMat();
+        Mat mIds = markerIds.getMat();
```
Comment (vrabaud) [inline modules/objdetect/src/aruco/charuco_detector.cpp:30]: BTW, you can keep the const here too.

File: modules/objdetect/src/aruco/charuco_detector.cpp [hunk 3]
```diff
@@ -27,10 +27,9 @@ struct CharucoDetector::CharucoDetectorImpl {
...
+        Mat chCorners = charucoCorners.getMat();
+        Mat chIds = charucoIds.getMat();
         const vector<int>& boardIds = board.getIds();
 
         const vector<vector<int> >& nearestMarkerIdx = board.getNearestMarkerIdx();
```
[the remark this answers is not in the document]
Comment (asmorkalov) [reply modules/objdetect/src/aruco/charuco_detector.cpp]: Fixed.

**Review by AleksandrPanov — APPROVED** [review]
👍

**Comment by asmorkalov** [comment]
@vrabaud please take a look again.

**Review by vrabaud — APPROVED** [review]
Will the API be changed for `getNearestMarkerIdx` and such at some point to avoid a vector copy? Otherwise, this PR is fine.

**Pull request merged**